presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -32,9 +32,22 @@
  */
 public class AsyncQueue<T>
 {
-    private final LinkedBlockingQueue<Object> queue = new LinkedBlockingQueue<>();
+    private final LinkedBlockingQueue<T> queue = new LinkedBlockingQueue<>();
+    private final int maxQueueSize;
```

Comment:
Maybe name this `targetMaxQueueSize` or `softMaxQueueSize` since it's not a hard limit